app/services/hmpps_api/prison_api/offender_api.rb

```diff
@@ -29,32 +29,35 @@ def self.list(prison, page = 0, page_size: 20)
         booking_ids = data.map { |o| o.fetch('bookingId') }
         sentence_details = HmppsApi::PrisonApi::OffenderApi.get_bulk_sentence_details(booking_ids)
 
+        complexities = ComplexityMicroService.get_complexities offender_nos
         offenders = data.map do |payload|
           offender_no = payload.fetch('offenderNo')
-          HmppsApi::OffenderSummary.from_json(payload,
-                                              search_data.fetch(offender_no, {}),
-                                              latest_temp_movement: temp_movements[offender_no]).tap { |offender|
+          HmppsApi::OffenderSummary.new(payload,
+                                        search_data.fetch(offender_no, {}),
+                                        latest_temp_movement: temp_movements[offender_no],
+                                        complexity_level: complexities[offender_no]).tap { |offender|
             sentencing = sentence_details[offender.booking_id]
             offender.sentence = sentencing if sentencing.present?
           }
         end
         ApiPaginatedResponse.new(total_pages, offenders)
       end
 
-      def self.get_offender(offender_no)
+      def self.get_offender(raw_offender_no)
         # Bad NOMIS numbers mustn't produce invalid URLs
-        url_offender_no = URI.encode_www_form_component(offender_no)
-        route = "/prisoners/#{url_offender_no}"
+        offender_no = URI.encode_www_form_component(raw_offender_no)
+        route = "/prisoners/#{offender_no}"
         response = client.get(route)
 
         if response.empty?
           nil
         else
-          search_data = get_search_payload([url_offender_no])
-          temp_movements = HmppsApi::PrisonApi::MovementApi.latest_temp_movement_for([url_offender_no])
-          HmppsApi::Offender.from_json(response.first,
-                                       search_data.fetch(url_offender_no, {}),
-                                       latest_temp_movement: temp_movements[url_offender_no]).tap do |offender|
+          search_data = get_search_payload([offender_no])
+          temp_movements = HmppsApi::PrisonApi::MovementApi.latest_temp_movement_for([offender_no])
+          HmppsApi::Offender.new(response.first,
+                                 search_data.fetch(offender_no, {}),
+                                 latest_temp_movement: temp_movements[offender_no],
+                                 complexity_level: ComplexityMicroService.get_complexity(offender_no)).tap do |offender|
```

Comment:
Just pointing this out, but I don't have a perfect solution to it:

Taking this approach means we'll end up performing needless calls to the Complexity API for **most offenders in the service** – since male offenders don't have a Complexity of Need level. It seems a bit wasteful.

Let's not forget the fact there will be a (possibly very small) performance hit for performing these calls – it's far cheaper to simply not make the call at all when we know it's not needed.

Should this class have knowledge of which conditions must be met for a Complexity of Need level to be needed? It feels wrong to do that – but it'd be great to avoid these extra API calls if at all possible.

Reply:
I think we're stuffed here - as we don't really know which prison we're in. I suppose we could call PrisonService.womens_prison? on the prison_id returned...